daemon/libnetwork/drivers/bridge/port_mapping_linux.go

```diff
@@ -365,7 +476,22 @@ func (n *bridgeNetwork) reapplyPerPortIptables() {
 		}
 	}
 
-	if err := n.firewallerNetwork.AddPorts(context.Background(), mergeChildHostIPs(allPBs)); err != nil {
+	if err := n.firewallerNetwork.AddPorts(context.Background(), collectFirewallPorts(allPBs)); err != nil {
 		log.G(context.TODO()).Warnf("Failed to reconfigure NAT: %s", err)
 	}
 }
+
+// collectFirewallPorts collects all the types.PortBinding needed to reconfigure
+// the host firewall for a given list of port bindings.
+func collectFirewallPorts(pbs []portmapperapi.PortBinding) []types.PortBinding {
+	var fwdPorts, natPorts []types.PortBinding
+	for _, pb := range pbs {
+		if pb.Forwarding {
+			fwdPorts = append(fwdPorts, toFwdBinding(pb))
+		}
+		if pb.NAT.IsValid() {
+			natPorts = append(natPorts, toNATBinding(pb))
+		}
```

Comment:
NAT currently implies forwarding (the `nat` portmapper needs forwarding enabled if the proxy is disabled, but it doesn't set `Forwarding: true`, and I think giving the Firewaller two separate `PortBinding` objects for the same port would lead to duplicate-rule errors). So ...

```suggestion
		if pb.NAT.IsValid() {
			natPorts = append(natPorts, toNATBinding(pb))
		} else if pb.Forwarding {
			fwdPorts = append(fwdPorts, toFwdBinding(pb))
		}
```

If this is right, the godoc strings for the new `NAT`/`Forwarding` fields in the port binding will need an update.

(And could just use a single list?)

Reply:
I'm still not sure how portmappers should handle gw modes, or if they even need to be aware of it. I think that portmappers wouldn't need to worry about that if they could set both Forwarding and NAT at the same time if they support both.

But there'd be an edge-case: how should the bridge driver handle PBs with `Forwarding` set, but not `NAT`, while the network is `gw_mode=nat`? Either these PBs should be ignored, or the host firewall should be reconfigured to allow forwarding. I think the latter option makes more sense, but indeed we'd need to add refcounters to the nftabler.